teps/0098-workflows.md

```diff
@@ -256,11 +306,164 @@ Why should this TEP _not_ be implemented?
+
+Cons:
+- Risks reinventing the wheel by redoing a lot of work done by [other projects](#prior-art)
+- We'll likely run into the same Triggers pain points other projects have
+- Takes longer to get to a usable project than making use of an existing working project
+
+### Design Workflows API with Conformance Spec
+
+In this solution, we would design a Workflows API and declare early on what it means to be Workflows conformant,
+rather than building extensibility mechanisms directly into the Workflows API.
+The conformance spec would be based on the work done by [existing projects](#prior-art).
+Platform builders would create their own implementations of the Workflows API to support customized
+connections to different SCMs or customized secrets storage. The Tekton team would maintain an implementation of the API that supports only
+the most commonly used SCMs.
+
+Pros:
+- App developers can run their Workflows on different platforms without having to rewrite them
+- Platform builders can also introduce specific features that make the most sense for their platform
+- Existing implementations of E2E CI/CD solutions build on top of Tekton can work to become Workflows conformant
+as the reference implementation evolves.
+
+Cons:
+- It will be difficult to come up with a Workflows conformance spec without first creating a reference implementation
+that supports several SCMs. It's unclear whether designing a good conformance spec is easier than designing a good
+API extensibility mechanism (and later on declaring a conformance spec).
+- Platform builders who want to change only part of the implementation will be forced to recreate the entire implementation.
+  - For example, a platform builder might want to add some custom steps to the process of connecting to a repo, or add their own backing
+  for secrets storage, without rewriting much of the project.
+
+### Adopt Pipelines as Code as Workflows Replacement
+
+In this solution, originally proposed in [#866](https://github.com/tektoncd/community/issues/883),
+Pipelines as Code would be transferred to the tektoncd organization,
+built and released as a Tekton project, and likely renamed to "Workflows", since the scope
+of the Workflows project is larger than storing Pipelines in version control.
+The Tekton community would maintain the project and shape its future to meet the goals of the Workflows project.
+The doc [Pipelines as Code -> Tekton Workflows](https://docs.google.com/document/d/1JAnXFkvs4aeciAJEsl7454e0J3ZHbL1yCPncMBHlTtM/)
+compares Pipelines as Code to the goals of the Workflows project.
+
+Suggested steps:
+1. Design and implement a solution for parameter substitution in Pipelines as Code that's Tekton conformant.
+(The current [syntax](https://pipelinesascode.com/docs/guide/authoringprs/#authoring-pipelineruns-in-tekton-directory)
+for variable expansion isn't Tekton conformant.)
```

Comment:
Great point! The only docs I'm aware of for Tekton conformance are this [api spec](https://github.com/tektoncd/pipeline/blob/main/docs/api-spec.md) and specifically this [section on params](https://github.com/tektoncd/pipeline/blob/main/docs/api-spec.md#param); however, it only details which api fields must be supported and not which param syntax tekton implementations are required to support. This would be a really useful thing to have as part of our API spec. FYI @XinruZhang since Xinru is working on a more up to date conformance spec.

I do think `$(tt.params.foo)` syntax is more in line with Pipelines param syntax than `{{ foo }}`. Would it be possible to replace `{{ foo }}` with something more Tekton-esque, like `$(context.event.repo_url)` or something like that?

Reply:
> Is Tekton triggers conformant to Tekton pipeline?

Great question!! Not for now and unknown for the future (likely not for the near future).

The conformance policy is defined for Tekton Pipeline service which takes stable Pipelines CRDs in, schedules and runs the workloads defined by those CRDs. Any vendor that wants to provide Tekton Pipeline Service needs to conform to the policy that's defined for those stable CRDs.

The plan is for the first step, we'd like to start with four Tekton primitives: Task, TaskRun, Pipeline and PipelineRun which have been promoted to V1.

Therefore, for triggers, it is not a Tekton Pipeline Service, and it produces Tekton primitives that will be taken by Tekton services, so it would be fine for Triggers to use different variable expansion syntax as long as the primitives it produces conform to the conformance policy we define for Tekton Pipelines.

Yes the $(tt.params.variable) syntax can be tedious, but I don't think that's a problem. Workflow is actually the very right project to ease the tedious. I would love to take quote a metaphor from @bobcatfish about the Tekton Pipelines:

Tekton Pipelines is more of plumbing instead of toilet – The Tekton Pipelines is more of underline infrastructure, let other works to improve the UX.  For instance – triggers, catalog etc.

> Tekton pipelines don't necessarily understand the $(tt.params.variable) syntax better than a {{ variable }} syntax?

Yeah it doesn't necessarily if we are designing it in the first place, but now we've heavily relied on the syntax, both developers and users are already familiar with it.

Reply:
Thanks for having me in the conversation! @lbernick!

Re @chmouel

> Is Tekton triggers conformant to Tekton pipeline?

Great question!! Not for now and unknown for the future (likely not for the near future).

The conformance policy is defined for Tekton Pipeline service which takes stable Pipelines CRDs in, schedules and runs the workloads defined by those CRDs. Any vendor that wants to provide Tekton Pipeline Service needs to conform to the policy that's defined for those stable CRDs.

The plan is for the first step, we'd like to start with four Tekton primitives: Task, TaskRun, Pipeline and PipelineRun which have been promoted to V1.

Therefore, for triggers, it is not a Tekton Pipeline Service, and it produces Tekton primitives that will be taken by Tekton services, so it would be fine for Triggers to use different variable expansion syntax as long as the primitives it produces conform to the conformance policy we define for Tekton Pipelines.

Yes the $(tt.params.variable) syntax can be tedious, but I don't think that's a problem. Workflow is actually the very right project to ease the tedious. I would love to take quote a metaphor from @bobcatfish about the Tekton Pipelines:

Tekton Pipelines is more of plumbing instead of toilet – The Tekton Pipelines is more of underline infrastructure, let other works to improve the UX.  For instance – triggers, catalog etc.

> Tekton pipelines don't necessarily understand the $(tt.params.variable) syntax better than a {{ variable }} syntax?

Yeah it doesn't necessarily if we are designing it in the first place, but now we've heavily relied on the syntax, both developers and users are already familiar with it.

Reply:
> Is Tekton triggers conformant to Tekton pipeline? because Tekton pipelines doesn't necessary understand better the $(tt.params.variable) syntax more than a {{ variable }} syntax?

The difference is that in Triggers, you use the $(tt.params) syntax in a TriggerTemplate that does understand that syntax. While in PAC, you are using {{ variable }} in a PipelineRun object which does not understand it. IMO calling it a pipelinerun when it contains syntax not supported in a regular pipelinerun is a bit confusing.

Reply:
Agreed with Dibyo, and thanks for suggesting improved wording here. I removed references to this syntax not being conformant, since we haven't defined this in our conformance spec, and added the point that a PipelineRun doesn't understand the syntax so it can't be used in Tekton Pipelines without modification.

Reply:
yes that's correct @dibyom. We do have some feature request to expand that `{{ variable }}` as a complete CEL expression, that understand the payload and do anything with it.

A benefit of having it like this is user doesn't need to have to embed a pipelinerun in another CRD (like done in workflow) and can just keep a simple PipelineRun as it is directly in their repository. This simplify things quite a bit.